### PR TITLE
Yoga build host image on arc

### PR DESCRIPTION
**Pull request by Alex-Welsh**
Still putting the final touches to this PR but everything should be working correctly now.
All images built [here](https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/8433465635)
latest test [here](https://github.com/stackhpc/stackhpc-kayobe-config/actions/workflows/overcloud-host-image-build.yml)

### CODE REVIEW

**Comment by Alex-Welsh** [comment]
Current issues:
Rocky 8 images are building as rocky 9
Rocky 9 images are failing to boot due to to an issue with cloud-init config not being parseable:
```
[   12.623671] cloud-init[966]: 2024-03-27 09:48:06,919 - util.py[WARNING]: Failed loading yaml blob. Invalid format at line 102 column 5: "while parsing a block mapping
[   12.626039] cloud-init[966]:   in "<unicode string>", line 102, column 5:
[   12.627229] cloud-init[966]:         renderers: ['network-manager']
[   12.628358] cloud-init[966]:         ^
[   12.629054] cloud-init[966]: expected <block end>, but found '<block mapping start>'
[   12.630460] cloud-init[966]:   in "<unicode string>", line 103, column 7:
[   12.631628] cloud-init[966]:           activators: ['network-manager']
```
I think that issue is caused by a whitespace error on [this line](https://github.com/stackhpc/stackhpc-image-elements/blob/841947d98ca2dedf3b3b8a463f23d5507ccbcb06/elements/rocky-container-stackhpc/containerfiles/9-stackhpc#L9C110-L9C111) of the containerfile
A fix is [in progress](https://github.com/stackhpc/stackhpc-image-elements/compare/master...fix-rocky-containerfile)

File: terraform/host-image-builder/README.rst [hunk 1]
```diff
@@ -0,0 +1,9 @@
+============================
```
Comment (markgoddard) [inline terraform/host-image-builder/README.rst]: Was it not possible to reuse the aio TF config?

File: terraform/host-image-builder/vm.tf [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  count = local.image_is_uuid ? 0 : 1
+}
+
+data "openstack_networking_subnet_v2" "network" {
+  name = var.host_image_builder_subnet
+}
+
+resource "openstack_compute_instance_v2" "kayobe-host-image-builder" {
+  name         = var.host_image_builder_name
+  flavor_name  = "en1.medium"
```
Comment (markgoddard) [inline terraform/host-image-builder/vm.tf]: host_image_builder_flavor

File: etc/kayobe/pulp-repo-versions.yml [hunk 3]
```diff
@@ -64,8 +64,8 @@ stackhpc_pulp_repo_rocky_9_3_crb_version: 20231215T005810
 stackhpc_pulp_repo_rocky_9_3_extras_version: 20231211T120328
 stackhpc_pulp_repo_rocky_9_3_highavailability_version: 20231214T005538
 stackhpc_pulp_repo_treasuredata_4_version: 20230903T003752
-stackhpc_pulp_repo_ubuntu_cloud_archive_version: 20231019T125502
-stackhpc_pulp_repo_ubuntu_focal_security_version: 20231018T165217
-stackhpc_pulp_repo_ubuntu_focal_version: 20231018T165217
-stackhpc_pulp_repo_ubuntu_jammy_security_version: 20231020T074329
-stackhpc_pulp_repo_ubuntu_jammy_version: 20231020T074329
+stackhpc_pulp_repo_ubuntu_cloud_archive_version: 20240327T041648
```
Comment (markgoddard) [inline etc/kayobe/pulp-repo-versions.yml]: We might want to control when we roll this out separately. I guess we don't necessarily need/want to update the images in this PR.

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+---
+# This playbook is desinged to be used by the overcloud-host-image-build.yml
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]: Have you considered using the Grammarly VS Code plugin?
```suggestion
# This playbook is designed to be used by the overcloud-host-image-build.yml
```

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    - name: Write out clouds.yml
+      copy:
+        content: "{{ lookup('ansible.builtin.env', 'CLOUDS_YAML') }}"
+        dest: clouds.yaml
+        mode: 0664
+
+    - name: Write out secure.yml
+      no_log: true
+      vars:
+        - os_secrets:
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]: nit: no need for a list here (although it does work)

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            clouds:
+              openstack:
+                auth:
+                  application_credential_id: "{{ lookup('ansible.builtin.env', 'OS_APPLICATION_CREDENTIAL_ID') }}"
+                  application_credential_secret: "{{ lookup('ansible.builtin.env', 'OS_APPLICATION_CREDENTIAL_SECRET') }}"
+      copy:
+        content: "{{ os_secrets | to_nice_yaml }}"
+        dest: secure.yaml
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]: It might be worth removing this guy at the end using a block with `always`, and possibly clouds.yaml too.

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    local_image_path: "/opt/kayobe/images/overcloud-{{ os_distribution }}-{{ os_release }}/overcloud-{{ os_distribution }}-{{ os_release }}.qcow2"
+    image_name: "overcloud-{{ os_distribution }}-{{ os_release }}"
+  tasks:
+    - name: Write out clouds.yml
+      copy:
+        content: "{{ lookup('ansible.builtin.env', 'CLOUDS_YAML') }}"
+        dest: clouds.yaml
+        mode: 0664
+
+    - name: Write out secure.yml
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]:
```suggestion
    - name: Write out secure.yaml
```

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+# This playbook is desinged to be used by the overcloud-host-image-build.yml
+# GitHub workflow to upload newly-built images to a development cloud for
+# testing and use in CI.
+- name: Upload an OS image to Glance
+  hosts: seed
+  vars:
+    local_image_path: "/opt/kayobe/images/overcloud-{{ os_distribution }}-{{ os_release }}/overcloud-{{ os_distribution }}-{{ os_release }}.qcow2"
+    image_name: "overcloud-{{ os_distribution }}-{{ os_release }}"
+  tasks:
+    - name: Write out clouds.yml
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]:
```suggestion
    - name: Write out clouds.yaml
```

File: etc/kayobe/ansible/openstack-host-image-upload.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        - os_secrets:
+            clouds:
+              openstack:
+                auth:
+                  application_credential_id: "{{ lookup('ansible.builtin.env', 'OS_APPLICATION_CREDENTIAL_ID') }}"
+                  application_credential_secret: "{{ lookup('ansible.builtin.env', 'OS_APPLICATION_CREDENTIAL_SECRET') }}"
+      copy:
+        content: "{{ os_secrets | to_nice_yaml }}"
+        dest: secure.yaml
+        mode: 0664
```
Comment (markgoddard) [inline etc/kayobe/ansible/openstack-host-image-upload.yml]: 0600 or 0660

File: .github/workflows/overcloud-host-image-build.yml [hunk 10]
```diff
@@ -35,21 +35,39 @@ on:
...
     steps:
-      - uses: actions/checkout@v4
+      - name: Install Package
+        uses: ConorMacBride/install-package@main
+        with:
+          apt: git unzip nodejs python3-pip python3-venv openssh-server openssh-client jq
+
+      - name: Start the SSH service
+        run: |
+          sudo /etc/init.d/ssh start
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:54]: Why is this required?

File: .github/workflows/overcloud-host-image-build.yml [hunk 11]
```diff
@@ -110,36 +96,133 @@ jobs:
...
+      - name: Manage SSH keys
+        run: |
+          mkdir -p ~/.ssh
+          touch ~/.ssh/authorized_keys
+          cat src/kayobe-config/terraform/host-image-builder/id_rsa.pub >> ~/.ssh/authorized_keys
+          cp src/kayobe-config/terraform/host-image-builder/id_rsa* ~/.ssh/
+
+      - name: Update inventory
+        run: |
+          cat > src/kayobe-config/etc/kayobe/environments/ci-builder/inventory/hosts << EOF
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Can we change this file to look like this out of the box instead?

File: .github/workflows/overcloud-host-image-build.yml [hunk 12]
```diff
@@ -110,36 +96,133 @@ jobs:
...
+
       - name: Bootstrap the control host
         run: |
           source venvs/kayobe/bin/activate &&
           source src/kayobe-config/kayobe-env --environment ci-builder &&
           kayobe control host bootstrap
 
-      - name: Configure the seed host
+      - name: Configure the seed host (Builder VM)
+        continue-on-error: true
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Why?

File: .github/workflows/overcloud-host-image-build.yml [hunk 13]
```diff
@@ -110,36 +96,133 @@ jobs:
...
-          kayobe seed host configure
+          kayobe seed host configure -e seed_bootstrap_user=rocky --skip-tags network
+
+      - name: Install dependencies
+        continue-on-error: true
+        run: |
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe seed host command run \
+          --command "sudo dnf config-manager --set-enabled crb && sudo dnf -y install epel-release && sudo dnf -y install zstd debootstrap kpartx cloud-init" --show-output
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:213]: What's crb for?

Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml:213]: Surprised we don't have cloud-init

File: .github/workflows/overcloud-host-image-build.yml [hunk 14]
```diff
@@ -110,36 +96,133 @@ jobs:
...
+      - name: Configure the seed host (Builder VM)
+        continue-on-error: true
         run: |
           source venvs/kayobe/bin/activate &&
           source src/kayobe-config/kayobe-env --environment ci-builder &&
-          kayobe seed host configure
+          kayobe seed host configure -e seed_bootstrap_user=rocky --skip-tags network
+
+      - name: Install dependencies
+        continue-on-error: true
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Why?

File: .github/workflows/overcloud-host-image-build.yml [hunk 15]
```diff
@@ -243,6 +342,16 @@ jobs:
           KAYOBE_VAULT_PASSWORD: ${{ secrets.KAYOBE_VAULT_PASSWORD }}
         if: inputs.rocky9
 
+      - name: Show last error logs
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:343]: Is this just for testing?

File: .github/workflows/overcloud-host-image-build.yml [hunk 16]
```diff
@@ -345,83 +470,27 @@ jobs:
...
-          path: |
-            /opt/kayobe/images/overcloud-ubuntu-jammy/overcloud-ubuntu-jammy.stdout
-            /opt/kayobe/images/overcloud-ubuntu-jammy/overcloud-ubuntu-jammy.stderr
-          retention-days: 7
-        if: steps.build_ubuntu_jammy.outcome == 'failure'
+      - name: Copy logs back
+        continue-on-error: true
+        run: |
+          mkdir logs
+          scp -r rocky@$(jq -r .access_ip_v4.value src/kayobe-config/etc/kayobe/environments/ci-builder/tf-outputs.yml):/opt/kayobe/images/*/*.std* ./logs/
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:489]: Does this need to specify the SSH key?

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: terraform/host-image-builder/README.rst (hunk 1 above)
Comment (Alex-Welsh) [reply terraform/host-image-builder/README.rst]: I have a feeling there used to be a conflict in an earlier iteration that doesn't exist any longer. I'll have a go at merging them.

File: etc/kayobe/pulp-repo-versions.yml (hunk 3 above)
Comment (Alex-Welsh) [reply etc/kayobe/pulp-repo-versions.yml]: Good point. I'll leave it in for now since CI is useful for determining whether I'm building valid images, but remove it later

File: .github/workflows/overcloud-host-image-build.yml (hunk 10 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:54]: I just couldn't get to SSH to the remote machine without it. I can dig out the error if you're interested

File: .github/workflows/overcloud-host-image-build.yml (hunk 12 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Old debugging artifact

File: .github/workflows/overcloud-host-image-build.yml (hunk 13 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:213]: CRB is a prerequisite for EPEL https://wiki.rockylinux.org/rocky/repo/#notes-on-epel
I might be able to reduce the requirements here but it would take a bit of time to test.

File: .github/workflows/overcloud-host-image-build.yml (hunk 15 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:343]: I initially added it for testing before I could copy back the full logs, but I was going to leave it in since it's very convenient to just check for obvious errors.

File: .github/workflows/overcloud-host-image-build.yml (hunk 16 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:489]: Nope, seems to work without

File: .github/workflows/overcloud-host-image-build.yml (hunk 11 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: My assumption was that it could interfere with container image builds. I'll kick one off and see what happens

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 9 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Done

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 4 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Done <sup><sup>with great shame</sup></sup>

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 5 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Feels wrong but I'll go with it

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 6 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Done

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 7 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Done

File: etc/kayobe/ansible/openstack-host-image-upload.yml (hunk 8 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/openstack-host-image-upload.yml]: Done

File: .github/workflows/overcloud-host-image-build.yml (hunk 12 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Removed

File: .github/workflows/overcloud-host-image-build.yml (hunk 14 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Old debugging artifact

File: .github/workflows/overcloud-host-image-build.yml (hunk 10 above)
Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml:54]: Would definitely like to know why.

File: .github/workflows/overcloud-host-image-build.yml (hunk 11 above)
Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml]: We write out the inventory in the container build job, so shouldn't cause a problem.

File: .github/workflows/overcloud-host-image-build.yml [hunk 17]
```diff
@@ -110,36 +96,141 @@ jobs:
...
+        working-directory: ${{ github.workspace }}/src/kayobe-config/terraform/aio
+
+      - name: Generate terraform.tfvars
+        run: |
+          cat << EOF > terraform.tfvars
+          ssh_public_key = "id_rsa.pub"
+          ssh_username = "rocky"
+          aio_vm_name = "skc-host-image-builder"
+          # Must be a Rocky Linux 9 host to successfully build all images
+          # This MUST NOT be an LVM image. It can cause confusing conficts with the built image.
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:123]: heh, I remember this one. It happens if the VG has the same name on the host. Tried to fix but was unable to.

File: .github/workflows/overcloud-host-image-build.yml [hunk 18]
```diff
@@ -110,36 +96,141 @@ jobs:
...
+          aio_vm_image = "Rocky-9-GenericCloud-Base-9.3-20231113.0.x86_64.qcow2"
+          aio_vm_flavor = "en1.medium"
+          aio_vm_network = "stackhpc-ci"
+          aio_vm_subnet = "stackhpc-ci"
+          aio_vm_interface = "eth0"
+          EOF
+        working-directory: ${{ github.workspace }}/src/kayobe-config/terraform/aio
+        env:
+          SSH_USERNAME: "id_rsa.pub"
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Not a username

File: .github/workflows/overcloud-host-image-build.yml [hunk 19]
```diff
@@ -110,36 +96,141 @@ jobs:
...
+          VM_NAME: "skc-host-image-builder-${{ github.run_id }}"
+          VM_IMAGE: 
+          VM_FLAVOR: 
+          VM_NETWORK: 
+          VM_SUBNET: 
+          VM_INTERFACE: ${{ inputs.vm_interface }}
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Unused

File: .github/workflows/overcloud-host-image-build.yml [hunk 20]
```diff
@@ -110,36 +96,141 @@ jobs:
...
+          VM_TAGS: '["skc-host-image-build", "ref=${{ github.ref_name }}"]'
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Unused

File: .github/workflows/overcloud-host-image-build.yml (hunk 13 above)
Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml:213]: Probably not worth the time

File: etc/kayobe/stackhpc-overcloud-dib.yml [hunk 21]
```diff
@@ -21,7 +21,7 @@ stackhpc_overcloud_dib_name: "deployment_image"
 
 # StackHPC overcloud DIB image elements.
 stackhpc_overcloud_dib_elements:
-  - "{{ os_distribution }}-{% if os_distribution == 'rocky' %}container-stackhpc{% else %}minimal{% endif %}"
```
Comment (markgoddard) [inline etc/kayobe/stackhpc-overcloud-dib.yml]: I'd consider this as a separate PR to fix RL9 builds

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: .github/workflows/overcloud-host-image-build.yml (hunk 17 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:123]: Drove me insane for a bit that one

File: .github/workflows/overcloud-host-image-build.yml (hunk 18 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: removed

File: .github/workflows/overcloud-host-image-build.yml (hunk 19 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: removed

File: etc/kayobe/stackhpc-overcloud-dib.yml (hunk 21 above)
Comment (Alex-Welsh) [reply etc/kayobe/stackhpc-overcloud-dib.yml]: Agreed. It's also not fully fixed yet and I don't want to sink more time into it now so I'll put that on hold.

File: .github/workflows/overcloud-host-image-build.yml (hunk 11 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Done

File: terraform/host-image-builder/vm.tf (hunk 2 above)
Comment (Alex-Welsh) [reply terraform/host-image-builder/vm.tf]: Removed, using AIO

File: terraform/host-image-builder/README.rst (hunk 1 above)
Comment (Alex-Welsh) [reply terraform/host-image-builder/README.rst]: Seems fine now

File: .github/workflows/overcloud-host-image-build.yml (hunk 20 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Removed

File: .github/workflows/overcloud-host-image-build.yml (hunk 10 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:54]: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/8467978144/job/23199884304#step:22:216
ssh-keyscan just failed to run

Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml:54]: Ah. Looks like we don't do a control host bootstrap in the container build job, which is why we didn't hit this. And the aio job uses kayobe-automation which does it's own SSH stuff

File: etc/kayobe/pulp-repo-versions.yml (hunk 3 above)
Comment (Alex-Welsh) [reply etc/kayobe/pulp-repo-versions.yml]: reverted

File: etc/kayobe/overcloud-dib.yml [hunk 22]
```diff
@@ -71,7 +71,7 @@ overcloud_dib_host_packages_extra:
 overcloud_dib_git_elements_extra:
   - repo: "https://github.com/stackhpc/stackhpc-image-elements"
     local: "{{ source_checkout_path }}/stackhpc-image-elements"
-    version: "v1.6.0"
+    version: "v1.6.1"
```
Comment (markgoddard) [inline etc/kayobe/overcloud-dib.yml:74]: Should this be in a separate change?

Comment (Alex-Welsh) [reply etc/kayobe/overcloud-dib.yml:74]: Builds flat out fail without it, so I'd argue it should remain part of this change

Comment (markgoddard) [reply etc/kayobe/overcloud-dib.yml:74]: Ack

**Review by markgoddard — APPROVED** [review]
